ivre/db/__init__.py

```diff
@@ -2406,78 +2406,108 @@ class DBFlow(DB):
-        else:
-            self.db_types["flow"]["postgresql"] = PostgresDBFlow
-            self.db_types["nmap"]["postgresql"] = PostgresDBNmap
-            self.db_types["passive"]["postgresql"] = PostgresDBPassive
-            self.db_types["view"]["postgresql"] = PostgresDBView
+        self._data = self.get_class("data")
+        return self._data
+
+    @property
+    def agent(self):
         try:
-            from ivre.db.sql.sqlite import SqliteDBPassive
-        except ImportError:
+            return self._agent
+        except AttributeError:
             pass
-        else:
-            self.db_types["passive"]["sqlite"] = SqliteDBPassive
-        if urls is None:
-            urls = {}
+        self._agent = self.get_class("agent")
+        return self._agent
+
+    @property
+    def flow(self):
         try:
-            from ivre.db.maxmind import MaxMindDBData
-        except ImportError:
+            return self._flow
+        except AttributeError:
             pass
-        else:
-            self.db_types["data"]["maxmind"] = MaxMindDBData
+        self._flow = self.get_class("flow")
+        return self._flow
+
+    @property
+    def view(self):
         try:
-            from ivre.db.http import HttpDBNmap, HttpDBView
-        except ImportError:
+            return self._view
+        except AttributeError:
             pass
-        else:
-            self.db_types["nmap"]["http"] = HttpDBNmap
-            self.db_types["view"]["http"] = HttpDBView
-        for datatype, dbtypes in viewitems(self.db_types):
-            specificurl = urls.get(datatype, url)
-            if specificurl is not None:
-                specificurl = urlparse(specificurl)
-                if specificurl.scheme in dbtypes:
-                    setattr(
-                        self,
-                        datatype,
-                        dbtypes[specificurl.scheme](specificurl))
-                    getattr(self, datatype).globaldb = self
+        self._view = self.get_class("view")
+        return self._view
+
+    def get_class(self, purpose):
+        url = self.urls.get(purpose, self.url)
+        if url is not None:
+            url = urlparse(url)
+            try:
+                modulename, classname = self.db_types[purpose][url.scheme]
+            except (KeyError, TypeError):
+                utils.LOGGER.error(
+                    'Cannot get database for %s from %s',
+                    purpose,
+                    url.geturl(),
+                    exc_info=True,
+                )
+                return None
+            try:
+                module = __import__('ivre.db.%s' % modulename).db
+            except ImportError:
+                utils.LOGGER.error(
+                    'Cannot import ivre.db.%s for %s',
+                    modulename,
+                    url.geturl(),
+                    exc_info=True,
+                )
+                return None
+            for submod in modulename.split('.'):
```

Comment:
I don't understand why this is necessary, but it is probably linked to line 2497. If `module` is a module object representing `ivre.db.sql.sqlite`, then line 2508 should work directly.

Reply:
Yep, that's the same question (and the same answer).